AsyncDisplayKit/Private/ASDisplayNode+UIViewBridge.mm

```diff
@@ -282,212 +288,222 @@ - (void)setNeedsDisplay
       [rasterizedContainerNode setNeedsDisplay];
     });
   } else {
-    // If not rasterized (and therefore we certainly have a view or layer),
-    // Send the message to the view/layer first, as scheduleNodeForDisplay may call -displayIfNeeded.
-    // Wrapped / synchronous nodes created with initWithView/LayerBlock: do not need scheduleNodeForDisplay,
-    // as they don't need to display in the working range at all - since at all times onscreen, one
-    // -setNeedsDisplay to the CALayer will result in a synchronous display in the next frame.
-
-    _messageToViewOrLayer(setNeedsDisplay);
-
-    BOOL nowDisplay = ASInterfaceStateIncludesDisplay(_interfaceState);
-    // FIXME: This should not need to recursively display, so create a non-recursive variant.
-    // The semantics of setNeedsDisplay (as defined by CALayer behavior) are not recursive.
-    if (_layer && !_flags.synchronous && nowDisplay && [self __implementsDisplay]) {
-      [ASDisplayNode scheduleNodeForRecursiveDisplay:self];
+    if (self.nodeLoaded) {
+      if (ASDisplayNodeThreadIsMain()) {
+        _messageToViewOrLayer(setNeedsDisplay);
+      } else {
+        if (!_pendingViewState.hasChanges) {
+          [ASPendingStateController.sharedInstance registerNode:self];
+        }
+        [self __setNeedsDisplay];
+        [_pendingViewState setNeedsDisplay];
+      }
+    } else {
+      [self __setNeedsDisplay];
     }
   }
 }
 
 - (void)setNeedsLayout
 {
-  _bridge_prologue;
-  [self __setNeedsLayout];
-  _messageToViewOrLayer(setNeedsLayout);
+  _bridge_prologue_write;
+  if (self.nodeLoaded) {
+    if (ASDisplayNodeThreadIsMain()) {
+       [self __setNeedsLayout];
+        _messageToViewOrLayer(setNeedsLayout);
+    } else {
+      if (!_pendingViewState.hasChanges) {
+        [ASPendingStateController.sharedInstance registerNode:self];
+      }
+      // NOTE: We will call [self __setNeedsLayout] just before we apply
```

Comment:
Hmm unfortunately, if we read from the pending state in those cases we'll miss any changes the user made to the underlying view/layer. It would be safe to call `__setNeedsLayout` just after applying the pending state, though, right?